pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/blocks/DataBlockBuilder.java

```diff
@@ -0,0 +1,406 @@
+            break;
+          // Multi-value column
+          case BOOLEAN_ARRAY:
+          case INT_ARRAY:
+            setColumn(rowBuilder, byteBuffer, (int[]) value);
+            break;
+          case TIMESTAMP_ARRAY:
+          case LONG_ARRAY:
+            // LONG_ARRAY type covers INT_ARRAY and LONG_ARRAY
+            if (value instanceof int[]) {
+              int[] ints = (int[]) value;
+              int length = ints.length;
+              long[] longs = new long[length];
+              ArrayCopyUtils.copy(ints, longs, length);
+              setColumn(rowBuilder, byteBuffer, longs);
+            } else {
+              setColumn(rowBuilder, byteBuffer, (long[]) value);
+            }
+            break;
+          case FLOAT_ARRAY:
+            setColumn(rowBuilder, byteBuffer, (float[]) value);
+            break;
+          case DOUBLE_ARRAY:
+            // DOUBLE_ARRAY type covers INT_ARRAY, LONG_ARRAY, FLOAT_ARRAY and DOUBLE_ARRAY
+            if (value instanceof int[]) {
+              int[] ints = (int[]) value;
+              int length = ints.length;
+              double[] doubles = new double[length];
+              ArrayCopyUtils.copy(ints, doubles, length);
+              setColumn(rowBuilder, byteBuffer, doubles);
+            } else if (value instanceof long[]) {
+              long[] longs = (long[]) value;
+              int length = longs.length;
+              double[] doubles = new double[length];
+              ArrayCopyUtils.copy(longs, doubles, length);
+              setColumn(rowBuilder, byteBuffer, doubles);
+            } else if (value instanceof float[]) {
+              float[] floats = (float[]) value;
+              int length = floats.length;
+              double[] doubles = new double[length];
+              ArrayCopyUtils.copy(floats, doubles, length);
+              setColumn(rowBuilder, byteBuffer, doubles);
+            } else {
+              setColumn(rowBuilder, byteBuffer, (double[]) value);
+            }
+            break;
+          case BYTES_ARRAY:
+          case STRING_ARRAY:
+            setColumn(rowBuilder, byteBuffer, i, (String[]) value);
+            break;
+          default:
+            throw new IllegalStateException(String.format(
+                "Unsupported data type: %s for column: %s", rowBuilder._columnDataType[i],
+                rowBuilder._dataSchema.getColumnName(i)));
+        }
+      }
+      rowBuilder._fixedSizeDataByteArrayOutputStream.write(byteBuffer.array(), 0, byteBuffer.position());
+    }
+    return buildRowBlock(rowBuilder);
+  }
+
+  public static ColumnarDataBlock buildFromColumns(List<Object[]> columns, DataSchema dataSchema)
+      throws IOException {
+    DataBlockBuilder columnarBuilder = new DataBlockBuilder(dataSchema, BaseDataBlock.Type.COLUMNAR);
+    for (int i = 0; i < columns.size(); i++) {
+      Object[] column = columns.get(i);
+      columnarBuilder._numRows = column.length;
```

Comment:
(nit) this can be done once outside the loop and we should ideally assert that TransferableBlock size (numRows) across all columns is same